tests/components/songpal/test_config_flow.py

```diff
@@ -0,0 +1,249 @@
+        "X_ScalarWebAPI_ServiceType"
+    ].append("videoScreen")
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_SSDP}, data=ssdp_data,
+    )
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "not_songpal_device"
+
+
+async def test_sddp_exist(hass):
+    """Test discovering existed device."""
+    _create_mock_config_entry(hass)
+    result = await hass.config_entries.flow.async_init(
+        DOMAIN, context={"source": SOURCE_SSDP}, data=SSDP_DATA,
+    )
+    assert result["type"] == RESULT_TYPE_ABORT
+    assert result["reason"] == "already_configured"
+
+
+async def test_user_exist(hass):
+    """Test user adding existed device."""
+    mocked_device = _create_mocked_device()
+    _create_mock_config_entry(hass)
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER}, data=CONF_DATA
+        )
+        assert result["type"] == RESULT_TYPE_ABORT
+        assert result["reason"] == "already_configured"
+
+    mocked_device.get_supported_methods.assert_called_once()
+    mocked_device.get_interface_information.assert_called_once()
+
+
+async def test_import_exist(hass):
+    """Test importing existed device."""
+    mocked_device = _create_mocked_device()
+    _create_mock_config_entry(hass)
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=CONF_DATA
+        )
+        assert result["type"] == RESULT_TYPE_ABORT
+        assert result["reason"] == "already_configured"
+
+    mocked_device.get_supported_methods.assert_called_once()
+    mocked_device.get_interface_information.assert_not_called()
+
+
+async def test_user_invalid(hass):
+    """Test using adding invalid config."""
+    mocked_device = _create_mocked_device(True)
+    _create_mock_config_entry(hass)
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_USER}, data=CONF_DATA
+        )
+        assert result["type"] == RESULT_TYPE_FORM
+        assert result["step_id"] == "user"
+        assert result["errors"] == {"base": "connection"}
+
+    mocked_device.get_supported_methods.assert_called_once()
+    mocked_device.get_interface_information.assert_not_called()
+
+
+async def test_import_invalid(hass):
+    """Test importing invalid config."""
+    mocked_device = _create_mocked_device(True)
+    _create_mock_config_entry(hass)
+
+    with _patch_config_flow_device(mocked_device):
+        result = await hass.config_entries.flow.async_init(
+            DOMAIN, context={"source": SOURCE_IMPORT}, data=CONF_DATA
+        )
+        assert result["type"] == RESULT_TYPE_ABORT
+        assert result["reason"] == "connection"
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/foreign-sub/home-assistant/pullRequest?prid=5462445)